### ISSUE TITLE
Teams and Orgs backend

### ISSUE THREAD

**Issue by zlavergne** [hotosm/tasking-manager#1531]
We are working on implementing a simple model for teams and organizations to satisfy our internal requirements. It's a simple design that we hope can be the base of future teams/orgs infrastructure. 

Our current design creates the following tables:

#### organizations
list of organizations

| Column | Type|
|--|--|
| id|int|
|name|varchar|
|visibility|ENUM|

#### org_admins
list of users who can make changes to organizations

|Column|Type
|-|-|
|user_id (FK) |int|
|org_id(FK)|int|

#### teams
list of teams and the org they belong to  (team doesn't have to belong to an org)

|Column|Type
|-|-|
|id |int|
|name|varchar|
|org_id (FK)|int|

#### team_members
list of users in a team

|Column|Type
|-|-|
|team_id (FK) |int|
|user_id (FK)|int|
|role*|ENUM|

\* Roles: `manager` or `editor`

#### projects
Change the current organization implementation

|Column|Type|
|-|-|
|~organization_tags~||
|organization_id (FK)| int|

This is a simple implementation that satisfies our internal requirements. With that, we don't plan on building out all the features of teams but rather create a good base on which the rest of the infrastructure can be built. 

I would appreciate any feedback, especially if anyone sees any blockers for iteration.

**Comment by xamanu** [comment]
Thanks @zlavergne for bringing this up. I would like to ping here some people to contribute with your thoughts: @kamicut, @JorgeMartinezG, @uwaiszaki.

**Comment by uwaiszaki** [comment]
@xamanu @zlavergne 
1) Would it be good to assign multiple organisations to a single team so that some special teams can be formed. Eg-> Team of some project managers who can edit any project data in case of emergency.
2) According to this plan a team can contribute to any project of the organisation. Letting them assign a team to the project can give better control over the projects.

**Comment by xamanu** [comment]
*How many organizations per team and project*

Following on @uwaiszaki first questions. I think we need to define, whether a team and project shall be associated with exactly one single organization? Or multiple? And is it valid to not have any connection? The current redesign input proposes to have exactly one one. Can we live with this simplicity for now?

*Roles on a team level*

I think for us it is the essential idea of teams to have validator and project manager roles based on teams. I do consider even to assume to have a team for all, which would allow us to have a much more consolidated permission handling, and we don’t have to assume two use cases of global and team validator and project manager roles.

**Comment by xamanu** [comment]
Thinking about your second comment, @uwaiszaki, I think indeed it makes sense in a team approach for projects rather than the organization, which anyway can be derived from the teams table.
Assuming one project, one team, one organization and together with entity relations it would look like this:

![tm-teams](https://user-images.githubusercontent.com/97706/57145667-3eb2f580-6db3-11e9-8514-8bff18928c14.jpg)
[Editable file for draw.io](https://github.com/hotosm/tasking-manager/files/3142401/tm-teams.zip)

**Comment by JorgeMartinezG** [comment]
@zlavergne I have some questions:

1. The organization _visibility_ term refers to visibility of the organization, or visibility of a project created by users that belong to the org?
2. Is possible to find a user within multiple teams/orgs? 
3. Also, a user could be admin on multiple orgs?
4. Are projects be visible at different levels? Pubiic (All users), team level, org level or private(just project creator)

**Comment by uwaiszaki** [comment]
>  Other options would be to have organizations to be assigned to projects, independently from teams. Or having several organizations per team or project, which would need more thinking.
 >
@xamanu Is this structure valid for assigning teams and organisations independently.
![teams1](https://user-images.githubusercontent.com/32809190/57185831-b63a6f00-6ef0-11e9-9678-d35ba5500ac9.png)

**Comment by ramyaragupathy** [comment]
While we start with one-to-one relationship between tables, practical implementation might lead us to think of one-to-many relationship in many of these. Outlining my thoughts for further discussion:

- Though a mapping campaign could be a joint effort of multiple organisations, a project technically belongs to just one organisation.
- Each organisation can have multiple admins/project managers, but there is only one author per project.
- I assume the `visibility` field controls visibility for - the organisation and the projects associated w/ the organisation. May be could be either `public` or `private` in which case only invited people can check out on organisation details and projects.
- Teams should be formed out of people invited to an organisation - so teams are internal to an organisation and there could be multiple teams within an org. 
- Within an organisation, mappers can belong to different  teams: eg: I could be in a team of `building mappers` and also `highway validators`.  Additionally a mapper could be invited to an organisation but still not be part of any of the team. 
- A project can have multiple teams mapped to it.
- A mapper could belong to different organisations and multiple teams within an organisation
- Within the tasking manager environment everyone should be given just one role - `mapper`. 
- Any mapper could create an organisation and automatically become the admin for the org.
- Other roles that we already have should be applicable at organisation level.

I am definitely seeing certain downsides to this approach and I'd like to hear from others how they're thinking through this is in the long run.

**Comment by zlavergne** [comment]
I'll try my best to address everyone's questions and comments. Please let me know if I miss anything!

@uwaiszaki 
>1. Would it be good to assign multiple organisations to a single team, this can also help in creating some special teams. Eg-> Team of some project managers who can edit any project data in case of emergency.

Rather than adding multiple organizations to a team, we have the ability of duplicating members of a team to a new team that can then be added to an organization. We did keep portability of a team between orgs and projects in mind when designing and I thought that could be added on in the future but wouldn't be blocked by this design. Do you agree?

>2. According to this plan a team can contribute to any project of the organization. Letting them assign a team to the project can give them better control over the projects.

I agree but we don't need it for our internal structure. I was actually thinking an ACL structure would be a good way to go forward with those types of permissions. 

My basic thoughts on that: teams and projects belong to an org independently and permissions are implemented with an ACL. I should have mentioned this in the original issue since it's what I was thinking would be at the core of future implementations. 

@xamanu This may also address a couple of your comments. Something like this would allow a teams to be more transferable between projects (even projects that belong to another org) since we aren't trying assigning a single team to a single project.

@xamanu 

>Roles on a team level

Depending on how the team-project relationship is implemented, I was thinking that that you could have a team of validators and team of mappers, PM's, etc and that team handled the permissions rather than relying on an additional user roles on each team (which I think is what you mention - just want to clarify)
Some benefits I see to this (I might repeat some of yours, sorry):
* a non-regular validator could be put on a validation team and only validate on projects where that team is assigned to validate
* it's a lot less work to control the whole team's permissions rather than each user
* (relate to ^) it could get hard for a user to keep track of their permissions on different teams

@JorgeMartinezG 

>1. The organization visibility term refers to visibility of the organization, or visibility of a project created by users that belong to the org?

This is talking about visibility of the org. Right now I think the teams and projects would inherit from the org visibility but I think there needs to be more discussion on that part.

>2. Is possible to find a user within multiple teams/orgs?
>3. Also, a user could be admin on multiple orgs?

Yes and yes!

>4. Are projects visible at different levels? Pubiic (All users), team level, org level or private(just project creator)

I'm glad you brought this up. Short answer, yes! Ideally an org could have projects that anyone can edit(public), publicly visible but only editable by team X(partially public), or private. 

@ramyaragupathy 

>* Though a mapping campaign could be a joint effort of multiple organisations, a project technically belongs to just one organisation.
* Each organisation can have multiple admins/project managers, but there is only one author per project.

Exactly what I was thinking.

>* Within an organisation, mappers can belong to different teams: eg: I could be in a team of building mappers and also highway validators. Additionally a mapper could be invited to an organisation but still not be part of any of the team.

possible with the ACL implementation I mention above

>* Within the tasking manager environment everyone should be given just one role - mapper.

I was actually thinking we should still have "global" roles. I hadn't thought about it much but I imagine we would want to maintain the ability to have a truly public project that doesn't require any permissions beyond what's currently baked into the TM.



I'll add a few more notes that I think may influence future implementations: 
* I think for a full implementation on this, there needs to be third tier for "Campaigns" which would be similar to a team in structure. In addition to an additional layer of permissions, this would allow for some large scale stats based on an ongoing campaign
* Could there be a benefit to teams being hierarchical? i.e. a parent team of `building team` has `building mappers` and `building validators` and the `building team` is assigned to  a project. This isn't very necessary with the current design but it may help with transferring teams to projects under another org.. Just a thought.


Thanks for all the great feedback!

**Comment by xamanu** [comment]
Thanks for all your input. This is a very productive conversation.

I want to recap. To me it seems we have a basic agreement on these **basic assumptions**:

- Organizations have one or more administrators.
- Organizations have none, one, or many teams.
- Users can be administrator of none, one or many organizations.
- Users can be part of none, one, or many teams.
- Teams have exactly one association to an organization. 
- Projects can be configured to be:
  - Public (everybody can map; everybody can validate)
  - Standard (everybody can map; validators of an associated team can validate)
   - Visible (everybody can see; associated teams can do any activities)
  - Archived (everybody can see; nobody can do any activities)
  - Draft (only project managers of the team/organization can see)

And then there are some **open questions** that I think we should clarify for us:

*1. Shall teams handle one role per team or one role per team membership?*

- Should teams consist of one role? / Should teams reflect roles under an organization (team for validators, team for mappers, …)?
XOR
- Should teams handle roles internally (one team can contain all kind of roles, many mappers, validators)?

*2. Are projects connected to teams? And how?*

- Shall or must projects have exactly one association to a team?
- Shall or must projects have one or more connected teams? 
- Can one project have no team connected to it?

*3. Are projects connected to organizations? And how?*

- Shall or must projects have exactly one association to an organization?
- Shall or must projects have a connections to organizations through the teams associated with it?
- Can one project have no organization connected to it?

*4. Are teams handled separately or do they replace the current “permission system”?*

- Do we want to support global roles?
- Can we have teams for everything? (including migration path)

**Comment by kamicut** [comment]
I'm going to write some thoughts on teams, summing up ideas from the February TM design meetings, and also pooling from Development Seed's ongoing work on our own "OSM Teams" which are a way to have portable teams across the OSM ecosystem (through OAuth). We want these teams to be compatible with TM teams, so I'm sharing what we've come up with.

## orgs
I want to push back on an assumption here, is the need for both organizations and teams. I'm not sure what the pattern is that organizations are solving, but it seems that having hierarchical teams is a solution we came up with during the TM talks. 

So if we had `TeamX` be the overarching team with a certain level of visibility and projects, then `TeamX/country` would be a subset of members only for that country with specific projects. Or `TeamX/country/validators` an even smaller subset.

The reason I'm bringing this up is to reduce the friction on new signups to understand the hierarchy before starting to map. If they join say the MissingMaps global team (and not an org, followed by a team), that's all they need to do so that they have access to projects. It's only after further specialization that we can filter their projects.

## visibility
For visibility, the way we've thought about it is that it's on the team level, similar to Facebook groups:
1. Public - searchable
2. Private - searchable but members not listed
3. Unlisted - cannot be found in search

## joining
A team can be either open or closed to new members joining.

## user permissions
A user can allow or deny to be added to new teams without a request to join

## permissions
For permissions we've thought of the following:
- The creator of a team is the first moderator of that team
- The admin can assign team moderators and invite mappers to a team
For simplification, we are using Whatsapp rules: if an admin leaves and there are no admins it goes to the next person chronologically.

## projects (TM specific)
I haven't thought about the visibility of projects within the tasking manager, only the team structure. It seems that teams should have projects with inherited visibility if created as part of a team. I do like @xamanu's configuration.

## roles (TM specific)
I think this can be handled with subsets like I've stated above.

N.B: https://github.com/developmentseed/osm-teams has some but not all of our code (specifically the OAuth bits which were pulled somewhere else). We want to open source this soon so if you are interested please let me know.

**Comment by xamanu** [comment]
Thanks for your input @kamicut! I think the teams of teams ideas is worth while thinking through it. And I tried to mock-up how a compatible structure to your proposal on the TM side could look like.

![tm-teams-of-teams](https://user-images.githubusercontent.com/97706/57436405-23227180-722f-11e9-9fef-03413cb9cd97.jpg)

[tm-teams-of-teams.zip](https://github.com/hotosm/tasking-manager/files/3160931/tm-teams-of-teams.zip) (you can edit using draw.io)

@zlavergne what do you think about hierarchical teams? Would this fit for your use case, too? Or are there any downsides that you can see from the perspective of your internal requirements?

----

On the TM side of things, I would like let us think about where the roles are getting applied:

- Team of roles (team of validators, as suggested by @kamicut and @ramyaragupathy)
- Global roles (in all teams being a validator based on the general role)
- Local team roles (validator in team x, normal mapper in team y)
- Project based team roles (for this project, team x is going to be the group of validators)

**Comment by JorgeMartinezG** [comment]
I think there is a problem when implementing teams of teams, and it is related to permissions. If ACL (https://www.ory.sh/docs/keto/engines/acl) is going to be implemented:
1. Permissions necessarily need to be set at a team (_team_id_ field from db) level, not user (_user_id_) level.
2. When checking permissions, this approach will be computationally expensive. For instance, if we have an user that belongs to a team which is two levels below the team with no parent_id, which is the organization, the server will need to check on each level of the hierarchy if is allowed to perform any action of the given resource. 

@xamanu related on the TM side of things, **team_members** table allows us granular roles. For instance, a user could be a validator in tem x, but could be an admin on team y. We could use the _function_ field to set different roles for a same user. I want to confirm if teams implementation is going to be a separate service. If yes, we can keep same user roles for TM forks that do not need teams feature.

Let me know your opinions

**Comment by kamicut** [comment]
Just to clarify two points:

1. On subteams: subteams are just regular teams that can inherit visiblity and other permissions at creation time. They get their own projects and behave independently; the only validation needed is that the list of members is a strict subset of its parent. This is a UI problem, not a permissions problem. This in my mind simplifies things so that there is no complex ACL going on. 

2. On external implementation of teams: Our approach with osm-teams is to create lists of users with simple roles (basically who is allowed to add members), and leave all other permissions to implementing apps. There are two possible ways the TM could integrate with osm-teams (if they so wish): a) by implementing the API calls to create / add / delete members and roles using the osm-teams API or b) creating a syncing or import mechanism. In both cases, the TM specific roles have to be created on the TM side, but TM has flexibility in how loosely or tightly coupled it wants to be to an external service.

**Comment by zlavergne** [comment]
Thanks for the detailed feedback @kamicut.


### Organizations
> I'm not sure what the pattern is that organizations are solving, but it seems that having hierarchical teams is a solution we came up with during the TM talks.

I think hierarchical teams are a great addition but they aren't necessary for our internal implementation. I do think though that orgs will be useful in the Tasking Manager. **Let's assume from here on that teams will be hierarchical.**

Having multiple nested teams with one master team acting as the org vs. having a separate `org` structure is mostly semantics since their purpose is the same. The benefit I see to `org`s is that they can function as an umbrella for `teams`, `projects`, and `campaigns` as well as a sort of "logical separation". Each of them (teams, projects, campaigns) will have their own relationships to each other but being able to group them all under a singular organization seems valuable. 

IMO it would be semantically confusing (both to developers and users trying to understand the structures) if there was a master team that had campaigns, projects and teams.
 
@xamanu how does that line up with how HOT organizes everything and what you were thinking?

An important distinction is that teams handle the permissions and orgs function as an umbrella structure to organize projects, teams, campaigns logically and in the future allow for stats at the highest level to be gathered.

> The reason I'm bringing this up is to reduce the friction on new signups to understand the hierarchy before starting to map. If they join say the MissingMaps global team (and not an org, followed by a team), that's all they need to do so that they have access to projects. It's only after further specialization that we can filter their projects.

I never thought they would need to join an org then team and other teams in succession. Rather that they would join or be invited to exactly whatever team(s) they wanted to. 

### Visibility 
> A team can be either open or closed to new members joining.

@kamicut would this be determined by the teams `visibility` or separate?


### Teams Service
It's great to see your `osm-teams` work! 

> b) creating a syncing or import mechanism. 

I think this is the better option. My initial thought on the process: you give the TM an **optional** OSM-Teams API key that polls the api and updates the TM DB. I think it's important to have the option to create a non-OSM-Teams team too. 

### @xamanu's questions
> 1. Shall teams handle one role per team or one role per team membership?

I think having roles at the team level would allow for easier modularity and easier transfer of teams. It's also a lot less work to change one role than change 20 different users' roles ;)

> 2. Are projects connected to teams? And how?
> * Shall or must projects have exactly one association to a team?

I definitely don't think they should be restricted to **one** team. Especially if the teams handle the roles. You would _have_ to be able to have more than one project<->team association.

> * Shall or must projects have one or more connected teams?
> * Can one project have no team connected to it?

My initial answer is no because I think to not deter new editors, they shouldn't be required to join a team to map a public project. But perhaps it's a matter of adding them to some _general_ team automatically but still maintaining the requirement on the backend to be related to a team. 

> Are projects connected to organizations? And how?
> * Shall or must projects have exactly one association to an organization?

I have similar feelings as I do above for teams. 
> * Shall or must projects have a connections to organizations through the teams associated with it?

I think this would just be handled through inheritance, right?

> * Can one project have no organization connected to it?

I think so. 

>4. Are teams handled separately or do they replace the current “permission system”?
> * Do we want to support global roles?

Yes. For the same reason I think projects shouldn't _have_ to be associated with a team. 

> * Can we have teams for everything? (including migration path)

Perhaps... I brought this up in my response about teams<->projects.

**Comment by xamanu** [comment]
Thank you all for your feedback and input! It looks like we are moving forward in defining the structure and I see a lot of agreements.

### TM database structure

For the HOT perspective we started writing up some [short user stories](https://docs.google.com/document/d/1EgAWb3LgkyWYJVmYutW9I43J4LXMmssZmNeaY-zvlM0/edit?usp=sharing) around different scenarios we see people using teams and organisations in the TM. We validated these user stories against the ideas. 

As we are discussing, the initial proposal by @zlavergne makes a lot of sense, and validating it against our user stories and derived requirements, we would like to propose to extend it with some more fields, we talked about here in the conversation. And we think about adding a `project_teams` table, because we require a bit more flexibility and on the same time being able to define explicitly which teams are going to be connected to a project (compare with user stories number 4 and 5). See the following entity relationship diagram:

![tm-teams-and-orgs (3)](https://user-images.githubusercontent.com/97706/57777751-195aab80-7712-11e9-84cc-71f112998d53.jpg)


##### Statements

I would like to express some of the contents of the diagram and some assumptions into human language. This might be helpful for more people to understand the diagram and it is a second path for us to make sure, we are talking the about the same details.

- Teams have exactly one association to an organization.
- Organizations have one or more administrators.
- Organizations have none, one, or many teams.
- Users can be administrator of none, one or many organizations.
- Users can be part of none, one, or many teams.
- Projects can have none, one, or many teams.
- Projects have exactly one association to an organization. 
- Organization administrators can create, modify and delete teams of their organization and manage their members.
- Projects assign teams to certain roles (teams `a` and `b` for validators, team `z` as project managers)

### osm-teams

Nice. We are all on the same page, about the desirable integration of OSM Teams, but to realize this an optional extension for the Tasking Manager. The implementation of the component that connects the Tasking Manager to the OSM teams api can follow one of the  "two possible ways the TM could integrate with osm-team" that @kamicut mentioned and can map the osm-teams structure to the TM database. There we will have to decide whether `organisations` are teams with the highest hierarchy, or whether `organisations` are something the TM handles internally.

**Comment by kamicut** [comment]
## team imports

Quick reply re:

> you give the TM an optional OSM-Teams API key that polls the api and updates the TM DB. I think it's important to have the option to create a non-OSM-Teams team too.

The way I'm designing the auth is that a user has *their* teams, and they assign public vs private. So users would have to have access tokens to pull their own teams that are private. This is how for example you can pull a private photo from instagram into facebook, and so facebook stores instagram scoped tokens for each user (integrated apps).

Given that an imported private team has a specific owner and permissions attached, @xamanu do you think this will map cleanly to a TM team? 

## visibility vs membership
> A team can be either open or closed to new members joining.
> > @kamicut would this be determined by the teams visibility or separate?

Visibility in my view is whether you can see the members of a team or not (public vs private vs unlisted), it's a privacy setting. Open vs Closed is about whether you need to approve members to join or not. I'm also using Facebook terminology here [Privacy Settings](https://www.facebook.com/help/220336891328465?helpref=about_content), [Membership Approval](https://www.facebook.com/help/214260548594688?helpref=faq_content)

**Comment by verdy-p** [comment]
for you info I discussed that by email (on the hot mailing list) and made the schema clearer to interpret graphically (all foreign keys are referencing table from left to right).

This starts from 11 tables (currently presented) with a normalization problem (yellow highlighting)

![Annotation 2019-06-13 235430](https://user-images.githubusercontent.com/1387035/59494223-03651580-8e8d-11e9-8210-24e57752fb1d.png)

Then merges two tables (campaign_projects and projects) to solve that normalization ambiguity.

![Annotation 2019-06-14 001835](https://user-images.githubusercontent.com/1387035/59494225-03651580-8e8d-11e9-86d6-3b5452ddcb0e.png)

Then merges two other ones (teams and admins) (with shortened table names and some additional fields for managing history/future planning of roles in teams).

![Annotation 2019-06-14 003219](https://user-images.githubusercontent.com/1387035/59494226-03fdac00-8e8d-11e9-9e7f-f567d5945601.png)

**Comment by verdy-p** [comment]
Annotated version of the last image (schema unchanged). Made a bit clearer by "moving" the "users" table to avoid the "spaghettis".

![Annotation 2019-06-14 102548](https://user-images.githubusercontent.com/1387035/59495136-fa754380-8e8e-11e9-81b5-de924c715c7f.png)

The yellow highlights show the comments for what I received in emails about the fact that teams are not necessarily attached to an organisation.

and for the additional fields that can be added to the merged "members" table (which merges the "organisation_admins" table and the "team_members" table into just one for all).

**Comment by verdy-p** [comment]
Note: I used MS access locally only because it allows fast simulation with real schemas (that can be tested with random data). I don't intend to propose you to use Access of course for this development. I don't care much about the symbology of the graphic schema, as it does not matter for this discussion (there are different conventions used, MS Access using an oversimplified presentation which unfortunately does not mark the cardinalities in both directions).
You can redraw the schema as you want with your prefered graphic tool and you are free to use any database engine that matches your goals.

**Comment by xamanu** [comment]
I want to report back some progress:

* The backend work from @zlavergne and @uwaiszaki has been merged into the code base #1858 
* The frontend and user flow has been designed by @hexiaok and we have a [click-dummy](https://projects.invisionapp.com/share/7PUM47HQ3H9#/screens) for you to check out.
